Mn nz join existing list

michellerenehey opened the pull request:
## Description

This pull request adds an additional feature to our application, which includes the ability for a user to search a pre-existing list. We implemented a form on the Home component, in which a user can type a three-word code. Upon clicking 'Join List', we created a function to query existing collections in our Firestore, and check if a list already exists. If it does, a user is redirected to that list, and if it does not, they are given an error message that the list does not exist. 

## Related Issue

<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->
Closes #5 

## Acceptance Criteria

<!-- Include AC from the Github issue -->
If a user **doesn’t** already have a token:
- [x] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [x] The input that accepts the list token has a semantic `label` element associated with its
- [x] The user can submit this form with both the mouse and the `Enter` key
- [x] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [x] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.
## Type of Changes

<!-- Put an `✓` for the applicable box: -->

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before

<!-- If UI feature, take provide screenshots -->
![Home page before issue 5 updates](https://user-images.githubusercontent.com/69729289/179865949-d528ca12-98db-4d60-bb97-aafee385b655.png)

### After

<!-- If UI feature, take provide screenshots -->
![Home page after issue 5 updates](https://user-images.githubusercontent.com/69729289/179865852-ac231162-3333-4e08-8bb8-2ddf19db4eeb.png)

## Testing Steps/ QA Criteria

1. Checkout the branch "MN-NZ-join-existing-list" and run `npm run`

2. If you have an already assigned token, you will see items in the `/list` page. When you press the Home button, you will be automatically redirected back to the `/list` page. This is the programmed behaviour of the app.

3. How to join a list: 

3a. You will need to clear the local storage in your browser. The following example is for Chrome. To clean the local storage, open developer tools in your browser (right-click and select `Inspect`), select `Application` tab and go to the `Local Storage` section on the left and right-click on the corresponding line in this section. Select `Clear` and refresh the page. The token will be gone and clicking the Home button will redirect you to `/` page.

![Clear local storage in devtools Chrome](https://user-images.githubusercontent.com/63440229/179890515-15153d50-e2c7-4d25-9463-a7f866711ab4.png)

3b. In the input box `List Name` type `my test list` and then press Submit using your mouse or hit the Enter key. You will be redirected to the `/list` page and the corresponding list will be displayed.

4. To be able to test the app behaviour when the user tries to join a non-existing list, you should clear the local storage first using the step 3a and then try entering a random combination of signs, for example `abc123`. You will not be redirected and a line notifying the user about the error will be displayed.

mira-kine commented on src/api/firebase.js:
```diff
@@ -20,7 +22,7 @@ const firebaseConfig = {
 
 const app = initializeApp(firebaseConfig);
-const db = getFirestore(app);
+export const db = getFirestore(app);
```
good call exporting db from here!

michellerenehey replied:
This was actually from a previous iteration! We originally built all logic in `Home.jsx` and exported what we needed - but we have since moved it over to `Firebase.js` so I don't think this needs to be exported anymore! I have deleted it.

lindseyindev replied:
Good work on the refactor! db should always remain in the firebase.js file and never need to be exported or imported. We want to isolate it to that file and use it with the api calls that need it and then import export those client-side Create, Read, Update Destroy api functions.

mira-kine commented on src/views/Home.jsx:
```diff
@@ -1,22 +1,58 @@
+		e.preventDefault();
+		setError(false);
+		const querySnapshot = await findToken(joinListName);
+		if (querySnapshot.size >= 1) {
```
oooh I've never used .size before. Can I ask why you chose .size over .length?

michellerenehey replied:
Oh yes! Great question! We tried to use .length, but the querySnapshot that is returned actually isn't an array! It's weird - it's a collection of objects, I think? The documentation says 'the number of documents can be determined via the `empty` and `size` properties.' (More in the docs [here](https://firebase.google.com/docs/reference/js/v8/firebase.firestore.QuerySnapshot) and [here](https://firebase.google.com/docs/reference/js/v8/firebase.firestore.QuerySnapshot#size).)

lindseyindev replied:
Great question Mira! Thanks for the explanation Michelle!

Trevor-Rezac replied:
Thanks Mira and Lindsey. Mira asked basically my same question in a much more direct way. Thank you Michelle for the links to resources. I think I need to better familiarize myself with the firebase docs.

mira-kine reviewed (approved):
Thank you for the clean and thorough QA Criteria! Works locally on my end, code is neat and easy to follow. I added some comments more because I'm curious on the code choices you made, otherwise looks good.

michellerenehey commented:
> Everything seems to work as intended & the testing instructions were super easy to follow and understand! I was able to see `my test list` as well as a list I generated on my own.
> 
> The only feedback I have is that it seems like the imports on line 5 of `Home.jsx` all appear to be unused. I think we can probably remove those for now, unless we are needing them for some reason?
> 
> Also, I was wondering if someone wouldn't mind explaining a little further exactly what is happening with the `findToken` function and how you knew what to do there, as well as in the `handleSubmit` function? I assumed the information was in the firebase docs but I think I am just not understanding their docs very well. Essentially, I am trying to understand how you reached the conclusion that you did. If that makes sense.

Oh good call re: the unused variables on `Home.jsx`! That was leftover from a previous solution. I've updated it. 

I'd be happy to talk through our function decisions! Our overall goal was to have a user enter a token (a list name) into an input field, and check to see if that token exists in our Firebase collection. If the token exists in Firebase, do something. If it doesn't, throw an error. 

So, we needed to figure out how to check that in Firebase! Nadia had great direction in working in our Firebase Firestore, and we ended up finding [this resource](https://firebase.google.com/docs/firestore/query-data/queries) in the documentation to guide our querying. We originally wrote all of this logic in `Home.jsx` (hence, the unused variables!), but decided that it was better to keep all Firebase logic in the `Firebase.js` file. So, we moved the logic over and created a function that we could call, `findToken` (which takes a user's input and queries the Firestore for it's existence). 

Because this function is called on a user interacting with a form (clicking 'join existing list'), we tucked the logic inside of a `handleSubmit` function in the `Home` component. That function calls `findToken`, takes in the list name that a user entered (we managed that in state as joinListName) as an argument, and returns a querySnapshot. If there is something in the snapshot, the user is redirected, otherwise they are fed an error. 

Happy to talk through any other questions!

lindseyindev commented on src/views/Home.jsx:
```diff
@@ -1,22 +1,57 @@
+		if (querySnapshot.size >= 1) {
+			setListToken(joinListName);
+			navigateTo('/list');
+		} else if (querySnapshot.empty) {
```
Nice work on using the firebase functionalities like .size and .empty!

lindseyindev commented on src/views/Home.jsx:
```diff
@@ -1,22 +1,57 @@
+				<form onSubmit={handleSubmit}>
+					<div className="list-name">
+						<label htmlFor="list-name">
+							List Name:{' '}
```
Great work with the form and the default submit events!

nickytonline replied:
Refactor: use CSS margin instead of a blank space to put space between the label and input.

lindseyindev commented on src/views/Home.jsx:
```diff
@@ -1,22 +1,57 @@
+							/>
+						</label>
+					</div>
+					{error && <p>That list does not exist.</p>}
```
As soon as I tested the non-existent list I was like "oooh that UX feedback message is nice how does that work?" Great implementation with the if/else in your handleSubmit!

lindseyindev reviewed (approved):
Thank you for the thorough testing steps and your work on this issue!! 

I don't have anything really to add other than we're seeing a lot of handleClick and handleSubmit functions so I really want to guide everyone towards descriptive function names as a way to document code functionality
Instead of handleSubmit something like handleJoinExistingList or handleExistingListLookup
and instead of handleClick -> handleCreateNewList

and this is really just to improve the speed of code reading and having code that is self-documenting and it will really help as you're adding more handleClick-esque functions

With my approval you're set to merge!

nickytonline commented on src/api/firebase.js:
```diff
@@ -98,3 +100,9 @@ export async function deleteItem() {
 }
+
+export async function findToken(listId) {
+	const q = query(collection(db, listId));
```
Refactor (nitpick): consider a more descriptive variable name.

nickytonline reviewed (approved):
Some suggestions, but looks great!

nickytonline reviewed (approved):
Some suggestions, but looks great!
